### PR TITLE
Custom style

### PR DESCRIPTION
**Pull request by tricknotes**
`Site#base_hue` を変更すると、サイトの雰囲気が変わるようにしました。
## TODO
- [x] Just works
- [x] テスト直す & 追加する
- [x] 同じファイルをなんども動的に生成しないようどこかにキャッシュする
- [x] 管理画面の登録フォームをいい感じにする
- [x] spectram の assets を管理しやすいように配置する

### CODE REVIEW

**Comment by tricknotes** [comment]
@mayuco 
`$base-hue` を設定してスタイルをカスタマイズできるようにしているんですが、mayuco さんが管理画面の入力方法をいい感じにする案をお持ちだと小耳にはさみました。

考えている方法を教えてもらうのでもいいですし、このブランチに直接 PR 送ってもらってもいいですが、どう進めるとよいでしょうか！？

File: app/views/admin/sites/_form.html.haml [hunk 1]
```diff
@@ -17,7 +17,12 @@
   = f.text_field :ad_slot
   = f.text_area :head_tag
   = f.text_area :promotion_tag
```
Comment (mayuco) [inline app/views/admin/sites/_form.html.haml]: したいことはできているのですが、上部でcol-sm-10を指定するようになっているため、inputを2つ横並びにしにくく、このように書いてしまっています。必要に応じて修正してほしいです。

File: config/routes.rb [hunk 2]
```diff
@@ -39,14 +39,14 @@
...
     resource :feed, only: %i(show), path: "feed", controller: "feed"
-
     resource :sitemaps, only: %i(show), path: "sitemap"
-
     resource :robots, only: %i(show), path: "robots"
 
     resources :fixed_pages, param: :slug, path: "/", only: :show
   end
 
+  resource :custom_css, only: %i(show), constraints: {fqdn: /[\w.]+/}, path: "custom_css/:fqdn-:digest.css", controller: "custom_css"
```
Comment (tricknotes) [inline config/routes.rb]: えー、ここはスペースなしがいいな…(というカスタマイズを探してこよう…)

File: app/controllers/custom_css_controller.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  def current_site
+    @current_site ||= Site.find_by!(fqdn: params[:fqdn])
+  end
+
+  def redirect_to_correct_css_url
+    redirect_to self.class.helpers.asset_path(current_site.css_location) unless current_site.custom_css_available?
+  end
+
+  def generate_custom_css(site)
```
Comment (tricknotes) [inline app/controllers/custom_css_controller.rb]: Don't mind :-)

File: app/views/admin/sites/_form.html.haml (hunk 1 above)
Comment (tricknotes) [inline app/views/admin/sites/_form.html.haml]: :muscle:

File: config/routes.rb [hunk 4]
```diff
@@ -39,14 +39,14 @@
...
     resource :feed, only: %i(show), path: "feed", controller: "feed"
-
     resource :sitemaps, only: %i(show), path: "sitemap"
-
     resource :robots, only: %i(show), path: "robots"
 
     resources :fixed_pages, param: :slug, path: "/", only: :show
   end
 
+  resource :custom_css, only: %i(show), constraints: {fqdn: /[\w.\-:]+/}, path: "custom_css/:fqdn-:digest.css", controller: "custom_css"
```
Comment (tricknotes) [inline config/routes.rb]: We prefer missing space.
This rule should be changed :-)

**Comment by tricknotes** [comment]
I think this became review ready.

File: test/controllers/custom_css_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    @request.env["devise.mapping"] = Devise.mappings[:user]
+
+    sign_in create(:user, :admin)
+  end
+
+  test "show should render css with Site#base_hue" do
+    site = create(:site, base_hue: 120)
+    get :show, fqdn: site.fqdn, digest: "some-values"
+
+    assert_response :success
```
Comment (tricknotes) [inline test/controllers/custom_css_controller_test.rb]: We should assert the content of css.

File: app/models/site.rb [hunk 6]
```diff
@@ -20,4 +25,22 @@ class Site < ActiveRecord::Base
...
+
+  def js_location
+    js_url.presence || "themes/default/application"
+  end
+
+  def css_location
+    asset_host = Rails.application.config.action_controller.asset_host
+
+    url_options = {}
+    url_options[:digest] = Digest::MD5.hexdigest(["site-css", base_hue, "layout-version"].join(":")) # TODO: Add layout-version API to layout gem. And use it.
```
Comment (tricknotes) [inline app/models/site.rb]: I'll work for it after merged :muscle:

**Comment by tricknotes** [comment]
This approach has performance problem:<
I'll try alternative way that compile assets on `Site#base_hue` changed.

**Comment by tricknotes** [comment]
review ready ~

**Comment by tricknotes** [comment]
I'll make a PR only migration to deploy safely when this PR is accepted.

**Comment by tricknotes** [comment]
I sent a new PR for migration: https://github.com/bm-sms/daimon-news/pull/476

File: app/views/admin/sites/_form.html.haml [hunk 7]
```diff
@@ -18,6 +18,12 @@
...
+  = f.form_group label: {text: Site.human_attribute_name(:theme_color)} do
```
Comment (okkez) [inline app/views/admin/sites/_form.html.haml]: Add help text using `help` option, please.
If you cannot describe in English, please describe in Japanese on Slack or esa.io.

File: config/application.rb [hunk 8]
```diff
@@ -29,6 +29,10 @@ class Application < Rails::Application
       g.scaffold_controller = :scaffold_controller
     end
 
+    config.assets.configure do |env|
+      env.append_path Rails.root.join("tmp/custom_css")
```
Comment (okkez) [inline config/application.rb]: Prefer `env.append_path(Rails.root.join("tmp/custom_css"))`

File: app/views/admin/sites/_form.html.haml (hunk 7 above)
Comment (tricknotes) [inline app/views/admin/sites/_form.html.haml]: What should we describe as `help` ?

File: config/application.rb (hunk 8 above)
Comment (tricknotes) [inline config/application.rb]: I see

File: app/views/admin/sites/_form.html.haml (hunk 7 above)
Comment (okkez) [inline app/views/admin/sites/_form.html.haml]: I'm not sure about "hue".
I think we should describe following:
- What is "hue"
- This block can change theme color by "base hue" (This block can change xxx or not yyy).

File: test/models/site_test.rb [hunk 9]
```diff
@@ -92,4 +92,32 @@ def test_validate(data)
...
+      assert site.custom_hue_css_url
+    end
+
+    test "without #base_hue" do
+      site = create(:site, base_hue: nil)
+
+      assert_nil site.custom_hue_css_url
+    end
+
+    test "update #base_hue" do
```
Comment (okkez) [inline test/models/site_test.rb:109]: We need test for `200` -> `300` or some other value not `nil`.

File: test/models/site_test.rb [hunk 10]
```diff
@@ -92,4 +92,32 @@ def test_validate(data)
       assert_equal(expected, site.valid?)
     end
   end
+
+  sub_test_case "Custome style" do
+    test "with #base_hue" do
+      site = create(:site, base_hue: 300)
+
+      assert site.custom_hue_css_url
```
Comment (okkez) [inline test/models/site_test.rb]: Do we need asserttion for `site.custom_hue_css_url` pattern?

File: test/models/site_test.rb (hunk 9 above)
Comment (tricknotes) [inline test/models/site_test.rb:109]: Ah. it's a good point.

File: test/models/site_test.rb (hunk 10 above)
Comment (tricknotes) [inline test/models/site_test.rb]: Yes, I'll do 🙆

File: test/models/site_test.rb [hunk 11]
```diff
@@ -92,4 +92,35 @@ def test_validate(data)
...
+      assert_valid_custom_css site
+      site.update!(base_hue: 300)
+      assert_valid_custom_css site
+      site.update!(base_hue: nil)
+      assert_nil site.custom_hue_css_url
+    end
+
+    def assert_valid_custom_css(site)
+      assert_match(/\.css\z/, site.custom_hue_css_url)
+      assert_match(/background-color:/, site.custom_hue_css.read)
```
Comment (tricknotes) [inline test/models/site_test.rb]: When it failed, it is too hard to correct css content or assertion.
ref: http://www.clear-code.com/blog/2012/1/18.html

File: test/models/site_test.rb [hunk 12]
```diff
@@ -92,4 +92,35 @@ def test_validate(data)
...
+    test "without #base_hue" do
+      site = create(:site, base_hue: nil)
+
+      assert_nil site.custom_hue_css_url
+    end
+
+    test "update #base_hue" do
+      site = create(:site, base_hue: nil)
+
+      assert_nil site.custom_hue_css_url
```
Comment (okkez) [inline test/models/site_test.rb]: Prefer `assert_nil(site.custom_hue_css_url)`

File: test/models/site_test.rb [hunk 13]
```diff
@@ -92,4 +92,35 @@ def test_validate(data)
...
+      site.update!(base_hue: 200)
+      assert_valid_custom_css site
```
Comment (okkez) [inline test/models/site_test.rb]: Prefer `assert_valid_custom_css(site)`

File: test/models/site_test.rb [hunk 14]
```diff
@@ -92,4 +92,35 @@ def test_validate(data)
...
+      site.update!(base_hue: 300)
+      assert_valid_custom_css site
```
Comment (okkez) [inline test/models/site_test.rb]: ditto.

File: test/models/site_test.rb [hunk 15]
```diff
@@ -92,4 +92,35 @@ def test_validate(data)
...
+      site.update!(base_hue: nil)
```
Comment (okkez) [inline test/models/site_test.rb]: We should assert that css file content is changed or not when `base_hue` is changed.

File: test/models/site_test.rb (hunk 14 above)
Comment (okkez) [inline test/models/site_test.rb]: We should assert that css file content is changed or not when base_hue is changed.

**Comment by okkez** [comment]
:dancers: LGTM

**Comment by okkez** [comment]
Need rebase :bow:

**Comment by tricknotes** [comment]
> Need rebase :bow:

Sir! Yes, sir!
